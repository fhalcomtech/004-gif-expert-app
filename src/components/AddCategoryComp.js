import React, { useState } from 'react'

export const AddCategoryComp = ({setCategories}) => {
    const [category, setCategory] = useState('');
    const [timer, setTimer] = useState(0);
    const [msg, setMsg] = useState('');
    const handlerChangeCategory = (e) => setCategory(e.target.value); 
    const handlerFormSubmit = (e) => {
        e.preventDefault();
        if(category.length<1) return;

        setCategories(curr => {
            if(timer){
                setTimer(tm => {
                    clearTimeout(tm);
                    return 0;
                });
            }
            if(curr.includes(category))
            {
                setMsg('La categoria ya existe')
                return curr;
            }
            else
            {
                setMsg('La categoria fue Agregada con exito')
                return ([category, ...curr]);
            }
        });
        setCategory('');
        setTimer(tm=>setTimeout(()=>{setMsg('')},5000));
        
    }

  return (
      <form onSubmit={handlerFormSubmit}>
          <h2>Gif Expert App</h2> 
          <h5>{msg}</h5>
          <input type='text' value={category} onChange={handlerChangeCategory}/>
          <button type='submit'>Save</button>
      </form>
  )
}
