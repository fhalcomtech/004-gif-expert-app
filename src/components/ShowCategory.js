import React from 'react'
import PropTypes from 'prop-types';

export const ShowCategory = ({ url, title}) => {
  return (
    <div className='card'>
        <div className='card-img'>
            <img src={url} alt={title}/>
        </div>    
        <div className='card-name'>
            {title.slice(0,10)}
        </div>
    </div>
  )
}

ShowCategory.propTypes = {
    url: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired
}
