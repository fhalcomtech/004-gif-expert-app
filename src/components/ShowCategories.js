import { useFetchGif } from '../hooks/useFetchGif';
import { ShowCategory } from './ShowCategory';

export const ShowCategories = ({category}) => {
   const {loading, data:images} = useFetchGif(category);
  return (
    <div className='category-container'>
        <h4 className='category-name'>
            {loading?'Loading...':category} 
        </h4>
        <div className='category-images'>
            {images.map(({id, url, title}) => (<ShowCategory key={id} url={url} title={title}/>))}
        </div>
    </div>
  )
}
