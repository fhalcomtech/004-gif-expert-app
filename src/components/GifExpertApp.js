import React, { useState } from 'react';
import { AddCategoryComp } from './AddCategoryComp';
import { ShowCategories } from './ShowCategories';

const GifExpertApp = ({defaultCategories= []}) =>
{
  const [categ, setCateg] = useState(defaultCategories);
  return (
    <>
      <AddCategoryComp setCategories={setCateg} />
      <hr/>
      <div>
         <ul>{categ.map(ct=><ShowCategories key={ct} category={ct} />)}</ul>
      </div>
    </>
  )

  
}

export default GifExpertApp;