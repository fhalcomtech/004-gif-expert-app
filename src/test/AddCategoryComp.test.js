import {shallow} from 'enzyme'
import '@testing-library/jest-dom';
import { AddCategoryComp } from '../components/AddCategoryComp';

describe("Create describe <AddCategoryComp/>", () => {
    const categs = ['goku','vegeta','itachi','naruto'];
    const getRandomArbitrary = (min, max) => (Math.random() * (max - min) + min)
    const fnGetCateg = () => categs.at(getRandomArbitrary(0,4));

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategoryComp setCategories={setCategories}/>);
    const inputText = wrapper.find(`input`);
    const formSim = wrapper.find("form");
    let varTests = "";

    beforeEach(()=>{
        jest.clearAllTimers();
        wrapper = shallow(<AddCategoryComp setCategories={setCategories}/>);
    });

    test("Creating Snapshop", ()=>{
        expect(wrapper).toMatchSnapshot();
    });

    test("Simulate Changes in input type text", ()=>{
        inputText.simulate("change",{target:{value:'Huguito'}});
    });

    test("Simulate form submit", ()=>{
        const submitMsg = 'Submit by form';
        formSim.simulate("submit",{preventDefault:()=>{varTests = 'Submit by form'}});
        expect(varTests).toBe(submitMsg);
    });

    test("Test call not call setCategories when category is empty" , (done)=>{
        const submitMsg = 'Submit by form';
        formSim.simulate("submit",{preventDefault:()=>{varTests = 'Submit by form'}});
        expect(varTests).toBe(submitMsg);
        expect(setCategories).not.toHaveBeenCalled();
        done();
    });


    test('Test call setCategories when category is filling out',(done)=>{
        const ctg = fnGetCateg();
        const inputCtg = wrapper.find('input');
        inputCtg.simulate('change',{target:{value:ctg}});
        wrapper.find('form').simulate('submit',{preventDefault:()=>{}});
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
        expect(inputCtg.props().value).toBe('');
        done();
    });
});