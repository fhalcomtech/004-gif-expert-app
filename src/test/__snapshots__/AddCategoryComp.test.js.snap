// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create describe <AddCategoryComp/> Creating Snapshop 1`] = `
<form
  onSubmit={[Function]}
>
  <h2>
    Gif Expert App
  </h2>
  <h5 />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button
    type="submit"
  >
    Save
  </button>
</form>
`;
